common/crypto.hpp

```diff
@@ -39,34 +41,67 @@ enum CryptoErrorCode {
 	VerificationError,
 };
 
+struct Args {
+	const string private_key_path;
+	const string private_key_passphrase;
+	const string ssl_engine;
+};
+
 class PrivateKey;
 using ExpectedPrivateKey = expected::expected<PrivateKey, error::Error>;
 
 class PrivateKey {
 public:
 	PrivateKey() {};
 
-	static ExpectedPrivateKey Load(const string &private_key_path) {
-		return Load(private_key_path, "");
+	static ExpectedPrivateKey Load(const Args &args);
+	static ExpectedPrivateKey LoadFromPEM(const string &private_key_path) {
+		return LoadFromPEM(private_key_path, "");
 	}
-	static ExpectedPrivateKey Load(const string &private_key_path, const string &passphrase);
 	static ExpectedPrivateKey LoadFromPEM(const string &private_key_path, const string &passphrase);
-	static ExpectedPrivateKey LoadFromPEM(const string &private_key_path);
+	static ExpectedPrivateKey LoadFromHSM(const Args &args);
 	static ExpectedPrivateKey Generate(const unsigned int bits, const unsigned int exponent);
 	static ExpectedPrivateKey Generate(const unsigned int bits) {
 		return PrivateKey::Generate(bits, MENDER_DEFAULT_RSA_EXPONENT);
 	};
 	error::Error SaveToPEM(const string &private_key_path);
+
+
 #ifdef MENDER_CRYPTO_OPENSSL
 	unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)> key {nullptr, [](EVP_PKEY *) { return; }};
 
 	EVP_PKEY *Get() {
 		return key.get();
 	}
 
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+	unique_ptr<ENGINE, void (*)(ENGINE *)> engine {nullptr, [](ENGINE *e) { return; }};
+#else
+	unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> default_provider {
+		nullptr, [](OSSL_PROVIDER *e) { return 0; }};
+	unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> hsm_provider {
+		nullptr, [](OSSL_PROVIDER *e) { return 0; }};
+#endif
+
 private:
 	PrivateKey(unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)> &&private_key) :
 		key(std::move(private_key)) {};
+
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+	PrivateKey(
+		unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)> &&private_key,
+		unique_ptr<ENGINE, void (*)(ENGINE *)> &&engine) :
+		key(std::move(private_key)),
+		engine {std::move(engine)} {};
+#else
+	PrivateKey(
+		unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)> &&private_key,
+		unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> &&default_provider,
+		unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> &&hsm_provider) :
+		key(std::move(private_key)),
+		default_provider {std::move(default_provider)},
+		hsm_provider {std::move(hsm_provider)} {};
```

Comment:
Can we have some alias around?
```
using PkeyPtr = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>;
using ProviderPtr = unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)>;
...
```

For being a header file, it became quite hard to read :eye:

Reply:
+1!

This should also help on reading the code where you use the constructors below.